File: mobile_app/lib/services/reviewer_service.dart
import 'api_service.dart';
import 'storage_service.dart';

/// Reviewer model for API responses
class Reviewer {
  final String id;
  final String name;
  final String email;
  final String role;
  final String expertise;
  final String? avatarUrl;
  final String status; // 'joined', 'pending', 'declined'
  final DateTime? joinedDate;
  final DateTime? invitedDate;

  Reviewer({
    required this.id,
    required this.name,
    required this.email,
    required this.role,
    required this.expertise,
    this.avatarUrl,
    required this.status,
    this.joinedDate,
    this.invitedDate,
  });

  /// Create Reviewer from JSON API response
  factory Reviewer.fromJson(Map<String, dynamic> json) {
    return Reviewer(
      id: json['id'] as String,
      name: json['name'] as String,
      email: json['email'] as String,
      role: json['role'] as String? ?? 'Reviewer',
      expertise: json['expertise'] as String? ?? 'Expert',
      avatarUrl: json['avatarUrl'] as String?,
      status: json['status'] as String,
      joinedDate: json['joinedDate'] != null
          ? DateTime.parse(json['joinedDate'] as String)
          : null,
      invitedDate: json['invitedDate'] != null
          ? DateTime.parse(json['invitedDate'] as String)
          : null,
    );
  }

  /// Convert to JSON
  Map<String, dynamic> toJson() => {
    'id': id,
    'name': name,
    'email': email,
    'role': role,
    'expertise': expertise,
    'avatarUrl': avatarUrl,
    'status': status,
    'joinedDate': joinedDate?.toIso8601String(),
    'invitedDate': invitedDate?.toIso8601String(),
  };
}

/// Service for managing idea reviewers
class ReviewerService {
  static final ReviewerService _instance = ReviewerService._internal();

  factory ReviewerService() {
    return _instance;
  }

  ReviewerService._internal();

  /// Get all reviewers for an idea
  /// 
  /// Returns a list of Reviewer objects (both active and pending)
  /// Throws Exception if API call fails
  Future<List<Reviewer>> fetchIdeaReviewers(String ideaId) async {
    try {
      final token = StorageService().getAccessToken();
      if (token == null) {
        throw Exception('No authentication token found');
      }

      final response = await ApiService.get(
        '/ideas/$ideaId/reviewers',
        token: token,
      );

      if (response is! Map<String, dynamic>) {
        throw Exception('Invalid response format');
      }

      final data = response['data'];
      if (data is! List) {
        throw Exception('Expected list of reviewers');
      }

      return data
          .map((json) => Reviewer.fromJson(json as Map<String, dynamic>))
          .toList();
    } catch (e) {
      throw Exception('Failed to fetch reviewers: $e');
    }
  }

  /// Remove a reviewer from an idea
  /// 
  /// Throws Exception if user is not authorized or reviewer not found
  Future<void> removeReviewer(String ideaId, String reviewerId) async {
    try {
      final token = StorageService().getAccessToken();
      if (token == null) {
        throw Exception('No authentication token found');
      }

      final response = await ApiService.delete(
        '/ideas/$ideaId/reviewers/$reviewerId',
        token: token,
      );

      if (response is! Map<String, dynamic>) {
        throw Exception('Invalid response format');
      }

      if (response['success'] != true) {
        throw Exception(response['details'] ?? 'Failed to remove reviewer');
      }
    } catch (e) {
      throw Exception('Failed to remove reviewer: $e');
    }
  }

  /// Resend invitation to a pending reviewer
  /// 
  /// Throws Exception if invitation is not pending or user is not authorized
  Future<void> resendInvitation(String ideaId, String reviewerId) async {
    try {
      final token = StorageService().getAccessToken();
      if (token == null) {
        throw Exception('No authentication token found');
      }

      final response = await ApiService.post(
        '/ideas/$ideaId/reviewers/$reviewerId/resend-invite',
        body: {},
        token: token,
      );

      if (response is! Map<String, dynamic>) {
        throw Exception('Invalid response format');
      }

      if (response['success'] != true) {
        throw Exception(response['details'] ?? 'Failed to resend invitation');
      }
    } catch (e) {
      throw Exception('Failed to resend invitation: $e');
    }
  }

  /// Cancel a pending invitation
  /// 
  /// Throws Exception if invitation is not pending or user is not authorized
  Future<void> cancelInvitation(String ideaId, String reviewerId) async {
    try {
      final token = StorageService().getAccessToken();
      if (token == null) {
        throw Exception('No authentication token found');
      }

      final response = await ApiService.delete(
        '/ideas/$ideaId/reviewers/$reviewerId/invite',
        token: token,
      );

      if (response is! Map<String, dynamic>) {
        throw Exception('Invalid response format');
      }

      if (response['success'] != true) {
        throw Exception(response['details'] ?? 'Failed to cancel invitation');
      }
    } catch (e) {
      throw Exception('Failed to cancel invitation: $e');
    }
  }

  /// Invite a reviewer to an idea
  /// 
  /// Throws Exception if user is not authorized or reviewer not found
  Future<Reviewer> inviteReviewer(String ideaId, String reviewerId) async {
    try {
      final token = StorageService().getAccessToken();
      if (token == null) {
        throw Exception('No authentication token found');
      }

      final response = await ApiService.post(
        '/ideas/$ideaId/reviewers',
        body: {'reviewerId': reviewerId},
        token: token,
      );

      if (response is! Map<String, dynamic>) {
        throw Exception('Invalid response format');
      }

      if (response['success'] != true) {
        throw Exception(response['details'] ?? 'Failed to invite reviewer');
      }

      return Reviewer.fromJson(response['data'] as Map<String, dynamic>);
    } catch (e) {
      throw Exception('Failed to invite reviewer: $e');
    }
  }

  /// Accept a reviewer invitation (user accepting their own invitation)
  /// 
  /// Throws Exception if invitation not found or user is not the reviewer
  Future<void> acceptInvitation(String ideaId, String reviewerId) async {
    try {
      final token = StorageService().getAccessToken();
      if (token == null) {
        throw Exception('No authentication token found');
      }

      final response = await ApiService.post(
        '/ideas/$ideaId/reviewers/$reviewerId/accept',
        body: {},
        token: token,
      );

      if (response is! Map<String, dynamic>) {
        throw Exception('Invalid response format');
      }

      if (response['success'] != true) {
        throw Exception(response['details'] ?? 'Failed to accept invitation');
      }
    } catch (e) {
      throw Exception('Failed to accept invitation: $e');
    }
  }

  /// Decline a reviewer invitation (user declining their own invitation)
  /// 
  /// Throws Exception if invitation not found or user is not the reviewer
  Future<void> declineInvitation(String ideaId, String reviewerId) async {
    try {
      final token = StorageService().getAccessToken();
      if (token == null) {
        throw Exception('No authentication token found');
      }

      final response = await ApiService.post(
        '/ideas/$ideaId/reviewers/$reviewerId/decline',
        body: {},
        token: token,
      );

      if (response is! Map<String, dynamic>) {
        throw Exception('Invalid response format');
      }

      if (response['success'] != true) {
        throw Exception(response['details'] ?? 'Failed to decline invitation');
      }
    } catch (e) {
      throw Exception('Failed to decline invitation: $e');
    }
  }
}

File: backend/src/services/reviewer.service.ts
import { prisma } from '../lib/prisma.ts';
import type { User, IdeaInvite } from '@prisma/client';

export const ReviewerService = {
  /**
   * Get all reviewers for an idea (both active and pending)
   */
  async getIdeaReviewers(ideaId: string) {
    // Get all invites for this idea (active = joined, pending = pending)
    const invites = await prisma.ideaInvite.findMany({
      where: { ideaId },
      include: {
        reviewer: {
          select: {
            id: true,
            name: true,
            email: true,
          },
        },
      },
      orderBy: { createdAt: 'desc' },
    });

    // Transform to match mobile app expectations
    return invites.map((invite) => ({
      id: invite.reviewer.id,
      name: invite.reviewer.name,
      email: invite.reviewer.email,
      role: 'Reviewer', // TODO: Add role field to IdeaInvite model if needed
      expertise: 'Expert', // TODO: Add expertise field to User profile
      status: invite.status.toLowerCase(),
      joinedDate: invite.status === 'ACCEPTED' ? invite.createdAt : null,
      invitedDate: invite.status === 'PENDING' ? invite.createdAt : null,
    }));
  },

  /**
   * Get a specific reviewer invite
   */
  async getReviewerInvite(ideaId: string, reviewerId: string) {
    return await prisma.ideaInvite.findUnique({
      where: {
        ideaId_reviewerId: { ideaId, reviewerId },
      },
      include: {
        reviewer: true,
      },
    });
  },

  /**
   * Check if user has access to manage reviewers (idea owner or organisation admin)
   */
  async canManageReviewers(ideaId: string, userId: string): Promise<boolean> {
    const idea = await prisma.idea.findUnique({
      where: { id: ideaId },
      select: {
        createdById: true,
        organisationId: true,
      },
    });

    if (!idea) {
      return false;
    }

    // Check if user is idea owner
    if (idea.createdById === userId) {
      return true;
    }

    // Check if user is organisation admin (if applicable)
    if (idea.organisationId) {
      const member = await prisma.membership.findUnique({
        where: {
          userId_organisationId: {
            userId,
            organisationId: idea.organisationId,
          },
        },
      });

      return member?.role === 'ADMIN';
    }

    return false;
  },

  /**
   * Remove a reviewer from an idea
   */
  async removeReviewer(ideaId: string, reviewerId: string) {
    const result = await prisma.ideaInvite.delete({
      where: {
        ideaId_reviewerId: { ideaId, reviewerId },
      },
    });

    return result;
  },

  /**
   * Resend invitation to a pending reviewer
   */
  async resendInvitation(ideaId: string, reviewerId: string) {
    // Verify the invite exists and is pending
    const invite = await prisma.ideaInvite.findUnique({
      where: {
        ideaId_reviewerId: { ideaId, reviewerId },
      },
    });

    if (!invite) {
      throw new Error('Reviewer invitation not found');
    }

    if (invite.status !== 'PENDING') {
      throw new Error('Can only resend pending invitations');
    }

    // Update the createdAt timestamp to mark as recently resent
    // In a real app, you'd send an email notification here
    const updated = await prisma.ideaInvite.update({
      where: {
        ideaId_reviewerId: { ideaId, reviewerId },
      },
      data: {
        createdAt: new Date(), // Mark as recently updated
      },
    });

    return updated;
  },

  /**
   * Cancel a pending invitation
   */
  async cancelInvitation(ideaId: string, reviewerId: string) {
    const invite = await prisma.ideaInvite.findUnique({
      where: {
        ideaId_reviewerId: { ideaId, reviewerId },
      },
    });

    if (!invite) {
      throw new Error('Reviewer invitation not found');
    }

    if (invite.status !== 'PENDING') {
      throw new Error('Can only cancel pending invitations');
    }

    return await prisma.ideaInvite.delete({
      where: {
        ideaId_reviewerId: { ideaId, reviewerId },
      },
    });
  },

  /**
   * Create a new reviewer invitation
   */
  async inviteReviewer(ideaId: string, reviewerId: string) {
    // Check if invite already exists
    const existing = await prisma.ideaInvite.findUnique({
      where: {
        ideaId_reviewerId: { ideaId, reviewerId },
      },
    });

    if (existing) {
      throw new Error('This user is already invited or reviewing this idea');
    }

    // Verify idea exists
    const idea = await prisma.idea.findUnique({
      where: { id: ideaId },
    });

    if (!idea) {
      throw new Error('Idea not found');
    }

    // Verify reviewer exists
    const reviewer = await prisma.user.findUnique({
      where: { id: reviewerId },
    });

    if (!reviewer) {
      throw new Error('Reviewer not found');
    }

    // Create invitation
    return await prisma.ideaInvite.create({
      data: {
        ideaId,
        reviewerId,
        status: 'PENDING',
      },
      include: {
        reviewer: {
          select: {
            id: true,
            name: true,
            email: true,
          },
        },
      },
    });
  },

  /**
   * Accept reviewer invitation
   */
  async acceptInvitation(ideaId: string, reviewerId: string) {
    const invite = await prisma.ideaInvite.findUnique({
      where: {
        ideaId_reviewerId: { ideaId, reviewerId },
      },
    });

    if (!invite) {
      throw new Error('Reviewer invitation not found');
    }

    return await prisma.ideaInvite.update({
      where: {
        ideaId_reviewerId: { ideaId, reviewerId },
      },
      data: {
        status: 'ACCEPTED',
      },
    });
  },

  /**
   * Decline reviewer invitation
   */
  async declineInvitation(ideaId: string, reviewerId: string) {
    const invite = await prisma.ideaInvite.findUnique({
      where: {
        ideaId_reviewerId: { ideaId, reviewerId },
      },
    });

    if (!invite) {
      throw new Error('Reviewer invitation not found');
    }

    return await prisma.ideaInvite.update({
      where: {
        ideaId_reviewerId: { ideaId, reviewerId },
      },
      data: {
        status: 'DECLINED',
      },
    });
  }
};
